File: backend/openmtscied/shared/crawlers/openstax_crawler.py
"""
OpenStax 爬虫包装器
"""

from .base_crawler import BaseCrawler
from . import register_crawler
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "scripts" / "scrapers"))


def crawl_openstax_wrapper(config: dict) -> dict:
    """OpenStax 爬虫包装器"""
    try:
        from openstax_scraper import OpenStaxScraper
        
        scraper = OpenStaxScraper()
        chapters = scraper.scrape_all_textbooks()
        
        output_file = config.get('output_file', 'data/textbook_library/openstax_chapters.json')
        scraper.save_chapters(chapters, output_file)
        
        return {
            "success": True,
            "total_items": len(scraper.textbooks),
            "scraped_items": len(chapters),
            "data": chapters
        }
    except Exception as e:
        return {
            "success": False,
            "total_items": 0,
            "scraped_items": 0,
            "error": str(e),
            "data": []
        }


register_crawler(
    crawler_id="openstax_textbooks",
    name="OpenStax Textbooks",
    handler=crawl_openstax_wrapper,
    description="爬取 OpenStax 教材章节"
)

File: backend-next/app/api/v1/admin/crawler/lib/openstax-crawler.ts
/**
 * OpenStax 爬虫
 * 从现有数据文件加载 OpenStax 教材章节
 */

import { registerCrawler } from './registry';
import type { CrawlerConfig, CrawlResult } from './types';
import { loadJsonFile, createSuccessResult, createErrorResult } from './utils';

async function crawlOpenStax(config: CrawlerConfig): Promise<CrawlResult> {
  try {
    console.log('[OpenStax] Starting crawler...');
    
    const outputFile = config.output_file || 'textbook_library/openstax_chapters.json';
    
    // 加载现有数据
    const data = loadJsonFile(outputFile);
    
    if (data.length === 0) {
      console.warn(`[OpenStax] No data found in ${outputFile}`);
      return createSuccessResult(0, 0, []);
    }
    
    console.log(`[OpenStax] Loaded ${data.length} chapters from ${outputFile}`);
    
    return createSuccessResult(data.length, data.length, data);
  } catch (error: unknown) {
    console.error('[OpenStax] Crawler error:', error);
    return createErrorResult(error.message || 'Unknown error');
  }
}

// 注册爬虫
registerCrawler(
  'openstax_textbooks',
  'OpenStax Textbooks',
  crawlOpenStax,
  '爬取 OpenStax 教材章节'
);

console.log('[OpenStax] Crawler registered');
